php-frontend/src/main/java/org/sonar/php/tree/symbols/DeclarationVisitor.java

```diff
@@ -107,10 +124,58 @@ public void visitClassDeclaration(ClassDeclarationTree tree) {
 
     IdentifierTree name = tree.name();
     SymbolQualifiedName qualifiedName = currentNamespace().resolve(name.text());
-    classSymbolDataByTree.put(tree, new ClassSymbolData(location(name), qualifiedName, superClassName, interfaceNames, tree.is(Tree.Kind.INTERFACE_DECLARATION)));
 
     symbolTable.declareTypeSymbol(tree.name(), globalScope, qualifiedName);
     super.visitClassDeclaration(tree);
+    ClassSymbolData classSymbolData = new ClassSymbolData(location(name), qualifiedName, superClassName, interfaceNames,
+      tree.is(Tree.Kind.INTERFACE_DECLARATION), methodsByClassTree.getOrDefault(tree, Collections.emptyList()));
+    classSymbolDataByTree.put(tree, classSymbolData);
+
+    currentClassTree = null;
+  }
+
+  @Override
+  public void visitAnonymousClass(AnonymousClassTree tree) {
+    boolean isInAnonymousClassBack = isInAnonymousClass;
+    isInAnonymousClass = true;
+    super.visitAnonymousClass(tree);
+    isInAnonymousClass = isInAnonymousClassBack;
+  }
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    if (currentClassTree == null) {
+      super.visitMethodDeclaration(tree);
+      return;
+    }
+
+    boolean backDidFindReturn = didFindReturn;
```

Comment:
The naming of this "back" variables is inconsistent. Here, it's a prefix, but it's a suffix for `isInAnonymousClassBack`.
An alternative would be to define such fields as lists or queues and push/pop elements.